### PR TITLE
 imsprog: init at 1.4.1

### PR DESCRIPTION
**Pull request by wucke13**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by wucke13** [comment]
@GrahamcOfBorg build imsprog

**Comment by wucke13** [comment]
@patka-123 Thank you for the review, I adjusted accordingly!
@GrahamcOfBorg build imsprog

**Review by patka-123 — APPROVED** [review]

**Comment by wucke13** [comment]
@GrahamcOfBorg build imsprog

**Comment by wucke13** [comment]
Done considering https://github.com/NixOS/nixpkgs/pull/291442. I opted against reusing the module, I think having a module which in the end adds one udev rule is not worth the extra file in the tree.

 @sund3RRR  Can you check this build on darwin? I see a chance that it might just work.
 @GrahamcOfBorg build imsprog

**Comment by sund3RRR** [comment]
> Done considering #291442. I opted against reusing the module, I think having a module which in the end adds one udev rule is not worth the extra file in the tree.
> 
> @sund3RRR Can you check this build on darwin? I see a chance that it might just work. @GrahamcOfBorg build imsprog

I don't know why NixOS can't enable udev rules from package without manually adding this in configuration. I think using it without a module is bad idea because new users may not understand why imsprog can't detect their programmer, because they don't know how to enable the rules.

I tested this thing on macOS and this doesn't work by imsprog design. This program needs additional feature to add macos device detection, so I set meta.platforms = linux for this package.

About your package: great job with autopatch hardcoded paths. But do you really need qt5.wrapQtAppsHook and qt5.qttools in buildInputs?

**Comment by wucke13** [comment]
> I don't know why NixOS can't enable udev rules from package without manually adding this in configuration.

Because adding packages to the store is an unprivileged action (available to all users), while adding udev rules is a privileged action (requiring elevated privileges, as one create privilege escalation schemes from modifying device access rules).

> I think using it without a module is bad idea because new users may not understand why imsprog can't detect their programmer, because they don't know how to enable the rules.

I believe it is fairly well understood among Nix users what a package can do, and what it can't do. And hence any user that used some non-standard peripheral (think UART transceiver, JTAG ports, ...) already knows the drill of plugging in their stuff to [services.udev.packages](https://search.nixos.org/options?channel=23.11&show=services.udev.packages&from=0&size=50&sort=relevance&type=packages&query=udev). However, there is a first time for everyone, and people trying this for the first time will most likely run into friction. I still prefer exposing users to the detail of "adding a udev rule to the system config" then to "enable the _some_ magical configuration via `programs.xyz.enable = true`. However, this is purely my own take on the matter.

> This program needs additional feature to add macos device detection

What would that be?

> About your package: great job with autopatch hardcoded paths. But do you really need qt5.wrapQtAppsHook and qt5.qttools in buildInputs?

Thank you for pointing this out. `wrapQtAppsHook` in `nativeBuildInputs` definitely is an oversight on my end. I'll `nix build .\#pkgsCross.aarch64-multiplatform.imsprog --print-build-logs` to find out if cross compilation works, that'll provide sufficient evidence whether qt5.qttools is required in `buildInputs` at all.

**Comment by sund3RRR** [comment]
> However, there is a first time for everyone, and people trying this for the first time will most likely run into friction. I still prefer exposing users to the detail of "adding a udev rule to the system config" then to "enable the _some_ magical configuration via `programs.xyz.enable = true`. However, this is purely my own take on the matter.

This is a huge problem of NixOS I think. This distro is really painful for new users. I want NixOS to become more popular, but it is impossible with current system design and attitude towards newbies.

> > This program needs additional feature to add macos device detection
> 
> What would that be?

Idk. I asked the darwin maintainers, they said that the program needs to detect devices darwin-way like. There is no udev rules on macOS.

File: pkgs/by-name/im/imsprog/package.nix
Comment (Pandapip1) [inline pkgs/by-name/im/imsprog/package.nix:10]:
```suggestion
    rev = "refs/tags/v${version}";
```
Including the full ref is often preferred.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Review by Pandapip1 — APPROVED** [review]
Diff LGTM

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Looks like this somehow got dropped

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
The commits should be squashed

**Comment by wucke13** [comment]
@Pandapip1 Done! I also bumped the version and added a missing `qt5.qtwayland` dep.

**Review by Pandapip1 — APPROVED** [review]
Diff LGTM

**Comment by wucke13** [comment]
@Lassulus would you mind taking a look if this can be merged?

File: pkgs/by-name/im/imsprog/package.nix
Comment (Aleksanaa) [inline pkgs/by-name/im/imsprog/package.nix:4]:
```suggestion
  pname = "imsprog";
```

File: pkgs/by-name/im/imsprog/package.nix
Comment (Aleksanaa) [inline pkgs/by-name/im/imsprog/package.nix:32]: Please move this comment above `postPatch`

File: pkgs/by-name/im/imsprog/package.nix
Comment (Aleksanaa) [inline pkgs/by-name/im/imsprog/package.nix:43]:
```suggestion
      --prefix PATH : "${lib.makeBinPath [ gnome.zenity wget ]}"
```

**Review by Aleksanaa — COMMENTED** [review]
Didn't test

**Comment by Lassulus** [comment]
Result of `nixpkgs-review pr 312900` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>imsprog</li>
  </ul>
</details>

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix:4]: Is it the norm to ditch the self given capitalization?

Comment (h7x4) [reply pkgs/by-name/im/imsprog/package.nix:4]: Not so much a norm, it's more of a convention (one of the bullet points further down): https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix]: @h7x4 thank you for the link! @Aleksanaa thank you for the feedback, I adjusted accordingly.

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix]: Done :smile:

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix]: Done :smile:

**Review by Pandapip1 — APPROVED** [review]

**Comment by Pandapip1** [comment]
Result of `nixpkgs-review pr 312900` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>imsprog</li>
  </ul>
</details>

**Review by Pandapip1 — APPROVED** [review]
Whoops didn't mean to dismiss my own review

File: pkgs/by-name/im/imsprog/package.nix
Comment (Aleksanaa) [inline pkgs/by-name/im/imsprog/package.nix:52]: Missing `mainProgram`

File: pkgs/by-name/im/imsprog/package.nix
Comment (Aleksanaa) [inline pkgs/by-name/im/imsprog/package.nix:42]: Don't use this as it will wrap the program twice. Use `qtWrapperArgs`: https://nixos.org/manual/nixpkgs/unstable/#qt-runtime-dependencies

Comment (sund3RRR) [reply pkgs/by-name/im/imsprog/package.nix:42]: `IMSProg_database_update` is a bash script, not a QT application

Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix:42]: To add to that: the last info box on https://nixos.org/manual/nixpkgs/stable/#qt-runtime-dependencies says:

> wrapQtAppsHook ignores files that are non-ELF executables.

However, an `ls` reveals that there is stil double-wrapping:

```console
$ ls -la result/bin/
.r-xr-xr-x 735k root  1 Jan  1970 .IMSProg-wrapped
.r-xr-xr-x 4,1k root  1 Jan  1970 .IMSProg_database_update-wrapped
.r-xr-xr-x  16k root  1 Jan  1970 .IMSProg_database_update-wrapped_
.r-xr-xr-x 160k root  1 Jan  1970 .IMSProg_editor-wrapped
.r-xr-xr-x  16k root  1 Jan  1970 IMSProg
.r-xr-xr-x  16k root  1 Jan  1970 IMSProg_database_update
.r-xr-xr-x  16k root  1 Jan  1970 IMSProg_editor
```

The culprit likely is that our new `wrap*AppsHook` implementation itself creates an ELF, that then gets wrapped by wrapQtAppsHook. Changing my custom wrapper from `preFixup` to `postFixup` solves the issue of double wrapping.

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix]: Thanks for the pointer, fixed with 5ce8db90a86534f5dadfdeeee466dfe4c4e6619c

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix:42]: Fixed in 5ce8db90a86534f5dadfdeeee466dfe4c4e6619c

File: pkgs/by-name/im/imsprog/package.nix
Comment (wucke13) [reply pkgs/by-name/im/imsprog/package.nix]: Thanks for the pointer, fixed with https://github.com/NixOS/nixpkgs/commit/5ce8db90a86534f5dadfdeeee466dfe4c4e6619c

**Comment by wucke13** [comment]
@Pandapip1 Thanks for the note, found and fixed :)

**Pull request merged**
